river/tree/_attribute_observer/numeric_attribute_class_observer_histogram.py

```diff
@@ -0,0 +1,109 @@
+import collections
+import functools
+
+from river import utils
+
+from .._attribute_test import NumericAttributeBinaryTest
+from .._attribute_test import AttributeSplitSuggestion
+from .attribute_observer import AttributeObserver
+
+
+class NumericAttributeClassObserverHistogram(AttributeObserver):
+    """Numeric attribute observer for classification tasks that discretizes features
+    using histograms.
+
+
+    Parameters
+    ----------
+    n_bins
+        The maximum number of bins in the histogram.
+    n_splits
+        The number of split points to evaluate when querying for the best split
+        candidate.
+    """
+
+    def __init__(self, n_bins: int = 256, n_splits: int = 32):
+        super().__init__()
+        self.n_bins = n_bins
+        self.n_splits = n_splits
+        self.hists = collections.defaultdict(functools.partial(utils.Histogram, max_bins=n_bins))
+
+    def update(self, att_val, class_val, sample_weight):
+        for _ in range(int(sample_weight)):
+            self.hists[class_val].update(att_val)
+
+        return self
+
+    def probability_of_attribute_value_given_class(self, att_val, class_val):
+        if class_val not in self.hists:
+            return 0.
+
+        total_weight = self.hists[class_val].n
+        for b in self.hists[class_val]:
+            if b.left <= att_val < b.right:
+                # Approximates the pdf of x by using the frequency in its corresponding
+                # histogram bin
+                return (b.count * (att_val - b.left) / (b.right - b.left)) / total_weight
+
+        return 0.
+
+    def best_evaluated_split_suggestion(self, criterion, pre_split_dist, att_idx, binary_only):
+        best_suggestion = None
+
+        low = min(h[0].right for h in self.hists.values())
+        high = min(h[-1].right for h in self.hists.values())
+
+        # If only one single value has been observed, then no split can be proposed
+        if low >= high:
+            return
+
+        n_thresholds = min(
+            self.n_splits,
+            max(map(len, self.hists.values())) - 1
+        )
+
+        thresholds = list(decimal_range(start=low, stop=high, num=n_thresholds))
```

Comment:
Why not return the list directly? Why implement it as a generator?

Reply:
That's a good question that I'll reserve for @MaxHalford, hahaha.

I simply leveraged the implementation from creme ;-)

Reply:
No reason in particular, apart from the fact that Python's `range` function is a generator.